### DISCUSSION TITLE
Spacing not affecting porosity or relative permeability curve

### DISCUSSION THREAD

**Discussion by Thraiye** [PMEAL/OpenPNM#2799] [Q&A] [answered]
Hello OpenPNM community,

I have been trying to observe the effect of different network parameters on porosity and relative permeability curves. Unfortunately, when changing the **spacing** value, there was **no impact on the porosity and relative permeability curves**. This was the case despite using values of significant variation i.e. 1e-6 vs 1e-3 metres for spacing. I first observed this while working with my created network, I later tried the simple example code available on the website, and I also observed the same.

It does not make sense to me why spacing does not have any impact. Can someone elaborate on this for me?

Thank you very much in advance for your assistance!

**Comment by ma-sadeghi** [comment]
I'm not aware of a priority curve. But could provide a script that reproduces tho behavior?

**Comment by ma-sadeghi** [comment]
Also, I think it might make sense for relative permeability curves to not change much. But first let me see the script

**Reply by Thraiye** [reply]
I used the example code for calculating relative permeability, where I only changed the spacing value for the second run:

# In[1]:


import numpy as np
import openpnm as op
import matplotlib.pyplot as plt
op.visualization.set_mpl_style()
np.random.seed(10)
get_ipython().run_line_magic('matplotlib', 'inline')
np.set_printoptions(precision=5)
op.Workspace().settings.loglevel = 40


# In[2]:


pn = op.network.Cubic(shape=[15, 15, 15], **spacing=1e-6**)   #for case 2 I changed spacing to 1e-3
pn.add_model_collection(op.models.collections.geometry.spheres_and_cylinders)
pn.regenerate_models()
air = op.phase.Air(network=pn,name='air')
air['pore.surface_tension'] = 0.072
air['pore.contact_angle'] = 180.0
air.add_model_collection(op.models.collections.phase.air)
air.add_model_collection(op.models.collections.physics.basic)
air.regenerate_models()
water = op.phase.Water(network=pn,name='water')
water.add_model_collection(op.models.collections.phase.water)
water.add_model_collection(op.models.collections.physics.basic)
water.regenerate_models()


# In[3]:


ip = op.algorithms.InvasionPercolation(network=pn, phase=air)
Finlets_init = pn.pores('left')
Finlets=([Finlets_init[x] for x in range(0, len(Finlets_init), 2)])
ip.set_inlet_BC(pores=Finlets)
ip.run()


# In[4]:


def sat_occ_update(network, nwp, wp, ip, i):
    r"""
        Calculates the saturation of each phase using the invasion
        sequence from either invasion percolation.
        Parameters
        ----------
        network: network
        nwp : phase
            non-wetting phase
        wp : phase
            wetting phase
        ip : IP
            invasion percolation (ran before calling this function)
        i: int
            The invasion_sequence limit for masking pores/throats that
            have already been invaded within this limit range. The
            saturation is found by adding the volume of pores and thorats
            that meet this sequence limit divided by the bulk volume.
    """
    pore_mask = ip['pore.invasion_sequence'] < i
    throat_mask = ip['throat.invasion_sequence'] < i
    sat_p = np.sum(network['pore.volume'][pore_mask])
    sat_t = np.sum(network['throat.volume'][throat_mask])
    sat1 = sat_p + sat_t
    bulk = network['pore.volume'].sum() + network['throat.volume'].sum()
    sat = sat1/bulk
    nwp['pore.occupancy'] = pore_mask
    nwp['throat.occupancy'] = throat_mask
    wp['throat.occupancy'] = 1-throat_mask
    wp['pore.occupancy'] = 1-pore_mask
    return sat


# In[5]:


def Rate_calc(network, phase, inlet, outlet, conductance):
    phase.regenerate_models()
    St_p = op.algorithms.StokesFlow(network=network, phase=phase)
    St_p.settings._update({'conductance' : conductance})
    St_p.set_value_BC(pores=inlet, values=1)
    St_p.set_value_BC(pores=outlet, values=0)
    St_p.run()
    val = np.abs(St_p.rate(pores=inlet, mode='group'))
    return val


# In[6]:


flow_in = pn.pores('left')
flow_out = pn.pores('right')


# In[7]:


model_mp_cond = op.models.physics.multiphase.conduit_conductance
air.add_model(model=model_mp_cond, propname='throat.conduit_hydraulic_conductance',
              throat_conductance='throat.hydraulic_conductance', mode='medium', regen_mode='deferred')
water.add_model(model=model_mp_cond, propname='throat.conduit_hydraulic_conductance',
              throat_conductance='throat.hydraulic_conductance', mode='medium', regen_mode='deferred')


# In[8]:


Snwp_num=10
flow_in = pn.pores('left')
flow_out = pn.pores('right')
max_seq = np.max([np.max(ip['pore.invasion_sequence']),
          np.max(ip['throat.invasion_sequence'])])
start = max_seq//Snwp_num
stop = max_seq
step = max_seq//Snwp_num
Snwparr = []
relperm_nwp = []
relperm_wp = []


# In[9]:


for i in range(start, stop, step):
    air.regenerate_models();
    water.regenerate_models();
    sat = sat_occ_update(network=pn, nwp=air, wp=water, ip=ip, i=i)
    Snwparr.append(sat)
    Rate_abs_nwp = Rate_calc(pn, air, flow_in, flow_out, conductance = 'throat.hydraulic_conductance')
    Rate_abs_wp = Rate_calc(pn, water, flow_in, flow_out, conductance = 'throat.hydraulic_conductance')
    Rate_enwp = Rate_calc(pn, air, flow_in, flow_out, conductance = 'throat.conduit_hydraulic_conductance')
    Rate_ewp = Rate_calc(pn, water, flow_in, flow_out, conductance = 'throat.conduit_hydraulic_conductance')
    relperm_nwp.append(Rate_enwp/Rate_abs_nwp)
    relperm_wp.append(Rate_ewp/Rate_abs_wp)


# In[10]:


plt.figure(figsize=[6,5])
plt.plot(Snwparr, relperm_nwp, '*-', label='Kr_nwp')
plt.plot(Snwparr, relperm_wp, 'o-', label='Kr_wp')
plt.xlabel('Snwp')
plt.ylabel('Kr')
plt.title('Relative Permeability in x direction')
plt.legend()


# In[11]:


#Calculating the porosity of the network
Vol_void = np.sum(pn['pore.volume'])+np.sum(pn['throat.volume'])
inlet = pn.pores('left')
outlet = pn.pores('right')
A = op.topotools.get_domain_area(pn, inlets=inlet, outlets=outlet)
L = op.topotools.get_domain_length(pn, inlets=inlet, outlets=outlet)
Vol_bulk = A * L
Poro = Vol_void / Vol_bulk
print(f'The value of Porosity is: {Poro:.2f}')

**Comment by Arenhart** [comment, marked as answer]
Hello, if I may interject, both porosity and relative permeability are adimensional properties and shouldn't change with spacing (unless you change the spacing to include anisotropy), as porosity is pore volume over total volume, and relative permeability is the phase permeability over the absolute permeability.
If you could try to check the network absolute permeability (https://openpnm.org/examples/applications/absolute_permeability.html), as I understand the subject, the property should be linearly proportional to the spacing.

**Comment by Thraiye** [comment]
Hello Arenhart; thank you for your response. 

**Spacing** is defined as the _distance between pore centres_ in each direction, so it is a parameter that will affect the total volume and not the pore volume. Given that the total volume is affected, the porosity should also change, I assume. 

Please help me clarify this if I am mistaken.

**Reply by Arenhart** [reply]
I'm sorry, my answer was incomplete, but it took some digging in the code to find the justifications.

It seems that when creating the actual spheres and cylinders for the PNM, the assigned pore sizes are a function of the largest possible sphere that can be assigned to each pore position, thus the total pore volume also changes. Basically, when changing the "spacing" input, you are scalling the whole network: boundaries, pores and throats. Again, if this is correct, porosity and relative permeability should be unchanged, while absolute permeability should scale linearly.

This is the code block that describes the pore size assignment:
https://github.com/PMEAL/OpenPNM/blob/b424204f114a86ed96e8ddb05d78c5788a4da695/openpnm/models/collections/geometry/spheres_and_cylinders.py#L4C1-L17C49
```
spheres_and_cylinders = {
    'pore.seed': {
        'model': mods.misc.random,
        'element': 'pore',
        'num_range': [0.2, 0.7],
        'seed': None,
    },
    'pore.max_size': {
        'model': mods.geometry.pore_size.largest_sphere,
        'iters': 10,
    },
    'pore.diameter': {
        'model': mods.misc.product,
        'props': ['pore.max_size', 'pore.seed'],
```

And here is part of the docstring of the largest_sphere function:
https://github.com/PMEAL/OpenPNM/blob/b424204f114a86ed96e8ddb05d78c5788a4da695/openpnm/models/geometry/pore_size/_funcs.py#L62
```
    Finds the maximum diameter pore that can be placed in each location without
    overlapping any neighbors.

    Parameters
    ----------
    %(network)s
    fixed_diameter : str
        Name of the dictionary key on ``network`` where the array containing
        pore diameter values is stored, if any. If not provided a starting
        value is assumed as half-way to the nearest neighbor.
```

So each sphere is given a diameter between 20% and 70% of the largest sphere possible in each site, and since the number of spheres is contant, this "largest sphere" is linearly proportional to the spacing chosen. I'm assuming throats have the same behaviour, resulting in the unchanging relative permeability.

Hope this helps.

Additionally, I wonder if this behaviour could be related to the issue #2800

**Reply by Thraiye** [reply]
Thank you very much Arenhart. This definitely helped explain it!
